File: COMP26120_u38174ys/lab3/python/bstree.py
import config

class bstree:
    def __init__(self, value):

        # initialize the tree
        self.verbose = config.verbose
        self.config = config
        self.value = value
        self.left = bstree()
        self.right = bstree()

    def get_hash(self, string):
            
            # get hash value for a string
            hash_value = 0
            for i in range(len(string)):
                hash_value = hash_value + ord(string[i])
            return hash_value

    def get_hash_code(self, hash_value):
                
                # get hash code for a hash value
                return hash_value % self.hash_table_size    

    

    def size(self):
        if (self.tree()):
            return 1 + self.left.size() + self.right.size()
        return 0
        
    def tree(self):
        # This counts as a tree if it has a field self.value
        # it should also have sub-trees self.left and self.right
        return hasattr(self, 'value')
        
    def insert(self, value):
        if (self.tree()):
            # TODO if tree is not NULL then insert into the correct sub-tree
            print("Placeholder - remove this print statement")
        else:
            # TODO otherwise create a new node containing the value
            print("Placeholder - remove this print statement")
        
    def find(self, value):
        if self.tree():
            # TODO complete the find function
            print("Placeholder - remove this print statement")
        return False
        
    # You can update this if you want
    def print_set_recursive(self, depth):
        if (self.tree()):
            for i in range(depth):
                print(" ", end='')
            print("%s" % self.value)
            self.left.print_set_recursive(depth + 1)
            self.right.print_set_recursive(depth + 1)
            
    # You can update this if you want
    def print_set(self):
        print("Tree:\n")
        self.print_set_recursive(0)
        
    def print_stats(self):
        # TODO update code to record and print statistic
        print("Placeholder - remove this print statement")
            
            

File: cwk/COMP26120_2021_b07683zz-lab5_solution/lab3/java/comp26120/bstree.java
package comp26120;

public class bstree implements set<String> {
    int verbose;

    String value;
    bstree left;
    bstree right;

    speller_config config;

    // TODO add fields for statistics
    int comparisons;
    int find_comparisons;
    int find = 0;

    public bstree(speller_config config) {
        verbose = config.verbose;
        this.config = config;
    }

    public int size() {
        // This presumes that if value is not null then (possibly empty) left and right
        // trees exist.
        if (tree()) {
            return 1 + left.size() + right.size();
        }
        return 0;
    }
    public int height() {
        if (tree()) {
            return 1 + left.height() > right.height() ? left.height() : right.height();
        }
        return 0;
    }

    public int comparisons(int type){
        if (tree()) {
            return type == 0? comparisons:find_comparisons + left.comparisons(type) + right.comparisons(type);
        }
        return type == 0? comparisons:find_comparisons;
    }

    public void insert(String value) {
        comparisons++;
        if (tree()) {
            // TODO if tree is not NULL then insert into the correct sub-tree
            if(value.equals(this.value)){
                return;
            }else if(value.compareTo(this.value) < 0){
                left.insert(value);
            }else{
                right.insert(value);
            }
        } else {
            // TODO otherwise create a new node containing the value and two sub-trees.
            this.value = value;
            left = new bstree(config);
            right = new bstree(config);
        }
    }

    public boolean find(String value) {
        find_comparisons++;
        if (tree()) {
            // TODO complete the find function
            if(value.equals(this.value)){
                find++;
                return true;
            }else if(value.compareTo(this.value) < 0){
                return left.find(value);
            }else{
                return right.find(value);
            }
        }
        // if tree is NULL then it contains no values
        if(size() == 1){
            find++;
        }
        return false;
    }

    private boolean tree() {
        return (value != null);
    }

    // You can update this if you want
    public void print_set_recursive(int depth) {
        if (tree()) {
            for (int i = 0; i < depth; i++) {
                System.out.print(" ");
            }
            System.out.format("%s\n", value);
            left.print_set_recursive(depth + 1);
            right.print_set_recursive(depth + 1);
        }
    }

    // You can update this if you want
    public void print_set() {
        System.out.print("Tree:\n");
        print_set_recursive(0);
    }

    public void print_stats() {
        System.out.format("bstree contains %d elements\n", size());
        System.out.format("bstree height is %d \n", height());
        System.out.format("average comparisons per insert is %.2f\n", (comparisons(0) + 0.0)/size());
        System.out.format("average comparisons per find is %.2f\n", (comparisons(1) + 0.0)/find);
    }
}
